hpl-2.2/man/man3/HPL_dlocmax.3
.TH HPL_dlocmax 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_dlocmax \- finds the maximum entry in matrix column.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&void\fR
\fB\&HPL_dlocmax(\fR
\fB\&HPL_T_panel *\fR
\fI\&PANEL\fR,
\fB\&const int\fR
\fI\&N\fR,
\fB\&const int\fR
\fI\&II\fR,
\fB\&const int\fR
\fI\&JJ\fR,
\fB\&double *\fR
\fI\&WORK\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_dlocmax\fR
finds  the maximum entry in the current column  and packs
the useful information in  WORK[0:3].  On exit,  WORK[0] contains the
local maximum  absolute value  scalar,  WORK[1] is the  corresponding
local row index,  WORK[2]  is the corresponding global row index, and
WORK[3] is the coordinate of the process owning this max.  When N  is
less than 1, the WORK[0:2] is initialized to zero, and WORK[3] is set
to the total number of process rows.
.SH ARGUMENTS
.TP 8
PANEL   (local input/output)    HPL_T_panel *
On entry,  PANEL  points to the data structure containing the
panel information.
.TP 8
N       (local input)           const int
On entry,  N specifies the local number of rows of the column
of A on which we operate.
.TP 8
II      (local input)           const int
On entry, II  specifies the row offset where the column to be
operated on starts with respect to the panel.
.TP 8
JJ      (local input)           const int
On entry, JJ  specifies the column offset where the column to
be operated on starts with respect to the panel.
.TP 8
WORK    (local workspace)       double *
On entry, WORK  is  a workarray of size at least 4.  On exit,
WORK[0] contains  the  local  maximum  absolute value scalar,
WORK[1] contains  the corresponding local row index,  WORK[2]
contains the corresponding global row index, and  WORK[3]  is
the coordinate of process owning this max.
.SH SEE ALSO
.BR HPL_dlocswpN \ (3),
.BR HPL_dlocswpT \ (3),
.BR HPL_pdmxswp \ (3),
.BR HPL_pdpancrN \ (3),
.BR HPL_pdpancrT \ (3),
.BR HPL_pdpanllN \ (3),
.BR HPL_pdpanllT \ (3),
.BR HPL_pdpanrlN \ (3),
.BR HPL_pdpanrlT \ (3),
.BR HPL_pdrpancrN \ (3),
.BR HPL_pdrpancrT \ (3),
.BR HPL_pdrpanllN \ (3),
.BR HPL_pdrpanllT \ (3),
.BR HPL_pdrpanrlN \ (3),
.BR HPL_pdrpanrlT \ (3),
.BR HPL_pdfact \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/pfact/HPL_dlocmax.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#ifdef STDC_HEADERS
void HPL_dlocmax
(
   HPL_T_panel *                    PANEL,
   const int                        N,
   const int                        II,
   const int                        JJ,
   double *                         WORK
)
#else
void HPL_dlocmax
( PANEL, N, II, JJ, WORK )
   HPL_T_panel *                    PANEL;
   const int                        N;
   const int                        II;
   const int                        JJ;
   double *                         WORK;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_dlocmax finds  the maximum entry in the current column  and packs
 * the useful information in  WORK[0:3].  On exit,  WORK[0] contains the
 * local maximum  absolute value  scalar,  WORK[1] is the  corresponding
 * local row index,  WORK[2]  is the corresponding global row index, and
 * WORK[3] is the coordinate of the process owning this max.  When N  is
 * less than 1, the WORK[0:2] is initialized to zero, and WORK[3] is set
 * to the total number of process rows.
 *
 * Arguments
 * =========
 *
 * PANEL   (local input/output)          HPL_T_panel *
 *         On entry,  PANEL  points to the data structure containing the
 *         panel information.
 *
 * N       (local input)                 const int
 *         On entry,  N specifies the local number of rows of the column
 *         of A on which we operate.
 *
 * II      (local input)                 const int
 *         On entry, II  specifies the row offset where the column to be
 *         operated on starts with respect to the panel.
 *
 * JJ      (local input)                 const int
 *         On entry, JJ  specifies the column offset where the column to
 *         be operated on starts with respect to the panel.
 *
 * WORK    (local workspace)             double *
 *         On entry, WORK  is  a workarray of size at least 4.  On exit,
 *         WORK[0] contains  the  local  maximum  absolute value scalar,
 *         WORK[1] contains  the corresponding local row index,  WORK[2]
 *         contains the corresponding global row index, and  WORK[3]  is
 *         the coordinate of process owning this max.
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
   double                     * A;
   int                        kk, igindx, ilindx, myrow, nb, nprow;
/* ..
 * .. Executable Statements ..
 */
   if( N > 0 )
   {
      A      = Mptr( PANEL->A, II, JJ, PANEL->lda );
      myrow  = PANEL->grid->myrow;
      nprow  = PANEL->grid->nprow;
      nb     = PANEL->nb;
      kk     = PANEL->ii + II + ( ilindx = HPL_idamax( N, A, 1 ) );
      Mindxl2g( igindx, kk, nb, nb, myrow, 0, nprow );
/*
 * WORK[0] := local maximum absolute value scalar,
 * WORK[1] := corresponding local  row index,
 * WORK[2] := corresponding global row index,
 * WORK[3] := coordinate of process owning this max.
 */
      WORK[0] = A[ilindx];         WORK[1] = (double)(ilindx);
      WORK[2] = (double)(igindx);  WORK[3] = (double)(myrow);
   }
   else
   {
/*
 * If I do not have any row of A, then set the coordinate of the process
 * (WORK[3]) owning this "ghost" row,  such that it  will never be used,
 * even if there are only zeros in the current column of A.
 */
      WORK[0] = WORK[1] = WORK[2] = HPL_rzero;
      WORK[3] = (double)(PANEL->grid->nprow);
   }
/*
 * End of HPL_dlocmax
 */
}
